mesonbuild/msetup.py

```diff
@@ -144,10 +150,15 @@ def add_vcs_ignore_files(self, build_dir: str) -> None:
         with open(os.path.join(build_dir, '.hgignore'), 'w', encoding='utf-8') as ofile:
             ofile.write(hg_ignore_file)
 
+    def add_cachedir_tag_file(self, build_dir: str) -> None:
```

Comment:
How about renaming add_vcs_ignore_files to add_ignore_files and having it write all 3 files?

Reply:
Well for some reason `add_vcs_ignore_files()` is guarded with
```
if not os.listdir(build_dir):
    self.add_vcs_ignore_files(build_dir)
    return src_dir, build_dir
```
in `validate_dirs()`, but I do not want to wait for a wipe build (which seems to be what this snippet is for) to add the CACHEDIR.TAG, as very few existing build directories would actually benefit from it. That wasn't the case before 89fe81835b190a897eaa484d82551b37cf2a6a01.

Reply:
> That wasn't the case before [89fe818](https://github.com/mesonbuild/meson/commit/89fe81835b190a897eaa484d82551b37cf2a6a01).

FWIW, it was already there: https://github.com/mesonbuild/meson/commit/89fe81835b190a897eaa484d82551b37cf2a6a01#diff-5ff3e6598a48c8bbe4eadd0316c1dbf401e67d23ce678f28f980d62a820788ccL157. The only difference is it got moved out of the function.

Reply:
AFAIK, it does that to not override if user edited/removed those files for any reason. In any case, as @tristan957 said you'll have to wipe when updating to a Meson version that supports CACHEDIR.TAG.

Reply:
> This wouldn't matter because when you update meson to a new minor version you have to wipe the build dir anyway.

Looking at the code, it seems like meson errors/wipes out only if the major version changed. It does not seem to happen for minor version, so it wouldn't wipe automatically any meson build folder to create the CACHEDIR.TAG. Maybe I'm just misunderstanding what you're referring to?

Reply:
This feature will only go in on a major version, so it doesn't matter.

Reply:
So I actually read the code wrong. I thought that [here](https://github.com/mesonbuild/meson/blob/90ce0841441506e3f409ab59ded1df8f2e6e7363/mesonbuild/coredata.py#L1090) the major version was the first number, but it uses `rsplit()` (reverse split) not the regular `split()`, so e.g. version `1.2.99` (current devel version) the major version number would be 2, not 1. I wondered why "major" version would be the first number, as it changed only once (it's a bit like GNOME's old 3.* prefix): now it makes sense. So I'm fine with making the change then.

Reply:
Well actually even then it does not work since the "major" version change only triggers a configuration regeneration, not a full wipe build, so the `if not os.listdir(build_dir):` still prevents the CACHEDIR.TAG from being created.

Reply:
It would be simpler to just drop this check altogether: I really can't see a reason to want to modify the .gitignore/.hgignore files, especially since the whole build directory should realistically not be committed.

Reply:
By major version, I was referring to the 2 in 1.2.0. Sorry for the miscommunication there.

Reply:
The 2 in 1.2.0 is a minor version, but not a bugfix version. Meson frequently releases minor versions -- also known as feature releases -- but major versions that have backwards-incompatible changes are a bit less common... (we've only had one major release so far and that was 1.0 which announced the stability policy to begin with).

Meson considers minor/feature releases to be fully backwards-compatible in terms of how meson works and the meson.build DSL, but the internal layout of the build directory's meson-private/ is, as strongly indicated, a private implementation detail and isn't stable except during bugfixes, which are by definition supposed to change nothing at all other than fixing errors. So that does require reconfiguring. A meson reconfigure will automatically detect that and print "Regenerating configuration from scratch" but it doesn't require wiping.